README.md
# vbExtra:
It is a component for Visual Basic 6 with extra objects, controls and functions.

Program URL: vbforums.com/showthread.php?865299


# Working with the source code:

"vbExtra (subclassing inside).vbp" and "vbExtra (subclassing outside).vbp", why two project files?

The ocx in the "bin" folder is compiled from "vbExtra (subclassing inside).vbp".
That ocx has no other dependency (it doesn't need any other file to work, beside of course the normal files that are already on Windows and the VB6 runtime).
But when working with the source code or when compiling the ocx, it is necessary to reference the typelib vbExTyp.tlb that is in the lib folder. This is not necessary in the client program.

The project "vbExtra (subclassing outside).vbp" is provided for development purposes, because while working on the component it is difficult when having the subclassing code within the same project in which we are working or debugging.
It needs a DLL, that is in the \Bin\ folder, named vbExSc1.dll.
If you compile the ocx from this project file [vbExtra (subclassing outside).vbp], then you'll need to also include that dependency file in your package.
But anyone that want to do changes or further development on the component can use the "vbExtra (subclassing outside).vbp" project for doing all the work, and once it is finished then use the "vbExtra (subclassing inside).vbp" to compile the final ocx.
Do not forget, if you added or removed files from the project (or changed their names), to manually edit and update the other vbp file accordingly.


# First release: 08/13/2018


Changelog.txt
03/Mar/2019 fixed a bug in PrinterEx
22/Feb/2019 Updated SSTabEx control
21/Feb/2019 Fixed a bug in FlexFn control
21/Feb/2019 Fixed a bug in PrinterEx
19/Jan/2019 Fixed a bug in DateEnter control (Change event wasn't fired when changed from a Date to a Null value)
15/Jan/2019 FlexFnObject updated
17/Dec/2018 fixed a bug in DateEnter control
16/Dec/2018 added another sample project for the ScrollableContainer control
16/Dec/2018 fixed a bug in ScrollableContainer control: the BottomFreeSpace property default setting was not working
30/Nov/2018 fixed a bug in DateEnter control on Windows XP
24/Nov/2018 fixed a bug in SSTabEx control
24/Nov/2018 updated FlexFn control
24/Nov/2018 fixed a bug in ScrollableContainer control
29/Oct/2018 SSTabEx control: enhanced/bug fixed highlight color for TabHoverHighlight and TabSelHighlight effects
16/Oct/2018 fixed some GDI memory resource leak issues
09/Oct/2018 added TabHoverEffect property to SSTabEx control
08/Oct/2018 added better support for multiple monitors set ups
08/Oct/2018 fixed a bug in PrintFn control
06/Oct/2018 fixed a bug in DateEnter control
06/Oct/2018 fixed a bug in ToolBarDA control
01/Oct/2018 fixed bugs in SSTabEx control in Sub SetThemeExtraData and Sub DrawTab in Windows 10
01/Oct/2018 fixed a bug in PrinterEx, Sub PrintRichTextBox. It didn't work if it was only one page.
26/Sep/2018 Chinese GUI captions update and small performance inprovement when printing FlexGrid
24/Sep/2018 better handle of mouse wheel and navigation of pages through the keyboard in the Print preview dialog
24/Sep/2018 added to FlexFn the property PrintCellsFormatting and a parameter with the same name to PrinterEx.PrintFlexGrid method that allows to ignore the (data) cells formatting when printing a FlexGrid
24/Sep/2018 added to PrinterEx, PrintFn and FlexFn the property PrintPrevUseOneToolBar that allows to have just one toolbar instead of two in the Print preview dialog
17/Sep/2018 fixed a bug in PrinterEx related to Tab positions and some other small fixes.
13/Sep/2018 fixed a bug in the PaintPicture method of Printer object
13/Sep/2018 fixed bugs in FlexFn
12/Sep/2018 fixed bugs and added functions to GridReportStyle's management
10/Sep/2018 fixed a bug in PaintPicture related to right margin
10/Sep/2018 added PrintFlexGrid method to PrinterEx
06/Sep/2018 added a feature that broke the binary compatibility so there is a version change. Now the call to PrinterEx.ShowPrintPreview can have parameters to send to the printing procedure.
04/Sep/2018 important update to PrinterEx object
31/Aug/2018 updated TrayIcon control
31/Aug/2018 updated DateEnter control
31/Aug/2018 updated SizeGrip control
31/Aug/2018 updated MouseWheelEnabler control
30/Aug/2018 updated PopupList control
29/Aug/2018 cGridHandler, FlexFn and FlexFnObject updated
28/Aug/2018 FlexFn: ptpFlexFnOtherFunctions was not configured
28/Aug/2018 cGridHandler: fixed a bug in Sub UpdateGridColumnsWidthsStretched
28/Aug/2018 Some minor fixes in ScrollableConatiner and changes in cGlobal.cls (and related files)
27/Aug/2018 Added custom message to ctlTrayIcon.ctl and cActivatePrevInstance.cls
26/Aug/2018 cPrinterEx: fixed a bug in PrintText function when CurrentX is out of the page ScaleWidth
25/Aug/2018 PrintFnObject: set Max(page) to the actual page count, and check that ToPage is not greater than the actual page count
25/Aug/2018 cPrinterEx: FromPage and ToPage were not set up in IPrinterEx_ShowPrintPreview
25/Aug/2018 CommonDialogExObject fixed a bug with cdlPDPageNums Flag
25/Aug/2018 cPrinterEx and PrintFn CommonDialogFlags property replaced by PrinterFlags and PageSetupFlags
25/Aug/2018 cPrinterEx: nFlags parameter in Sub IPrinterEx_ShowPrint and Sub IPrinterEx_ShowPageSetup wasn't taken into account.
23/Aug/2018 Changed folder names in the project: Bin ==> lib; Release ==> bin
17/Aug/2018 GUI localized captions for Chinese language updated
16/Aug/2018 Added more Unicode support for fonts names and CommonDialogEx
16/Aug/2018 SSTabEx code updated. Due to a VB6 bug, Single type numbers are localized in the properties saved (commas intead of points), so I decided to round them to integers when they are saved to avoid the problem.
15/Aug/2018 Added Unicode support for font names in PrinterEx
15/Aug/2018 Fixed FontCount property get in cPrinterEx
14/Aug/2018 Added Chinese Simplified GUI language support.

bas/mGridsFunctions.bas
Attribute VB_Name = "mGridsFunctions"
Option Explicit

Public Enum efnFlexSortSettings
    flexSortNone = 0
    flexSortGenericAscending = 1
    flexSortGenericDescending = 2
    flexSortNumericAscending = 3
    flexSortNumericDescending = 4
    flexSortStringNoCaseAscending = 5
    flexSortStringNoCaseDescending = 6
    flexSortStringAscending = 7
    flexSortStringDescending = 8
End Enum

Private mGridsArrowUpImageCollection As New Collection
Private mGridsArrowDownImageCollection As New Collection
Public gGridReportStyles As New Collection

Public Function GetGridArrowUpImage(nGrid As Object) As StdPicture
    On Error Resume Next
    Set GetGridArrowUpImage = mGridsArrowUpImageCollection(CStr(nGrid.BackColorFixed))
    On Error GoTo 0
    If GetGridArrowUpImage Is Nothing Then
        StoreGridArrowImages nGrid
    End If
    On Error Resume Next
    Set GetGridArrowUpImage = mGridsArrowUpImageCollection(CStr(nGrid.BackColorFixed))
    On Error GoTo 0
End Function

Public Function GetGridArrowDownImage(nGrid As Object) As StdPicture
    On Error Resume Next
    Set GetGridArrowDownImage = mGridsArrowDownImageCollection(CStr(nGrid.BackColorFixed))
    On Error GoTo 0
    If GetGridArrowDownImage Is Nothing Then
        StoreGridArrowImages nGrid
    End If
    On Error Resume Next
    Set GetGridArrowDownImage = mGridsArrowDownImageCollection(CStr(nGrid.BackColorFixed))
    On Error GoTo 0
End Function

Private Sub StoreGridArrowImages(nGrid As Object)
    Dim iPic As Control
    Dim iImage As StdPicture
        
    On Error Resume Next
    nGrid.Parent.Controls.Add "VB.Picturebox", "picAux1x"
    Set iPic = nGrid.Parent.Controls("picAux1x")
    On Error GoTo TheExit:
    
    iPic.Width = nGrid.Parent.ScaleX(17 + nGrid.GridLineWidth, vbPixels, nGrid.Parent.ScaleMode)
    iPic.Height = nGrid.Parent.ScaleY(11, vbPixels, nGrid.Parent.ScaleMode)
    iPic.BorderStyle = 0
    iPic.AutoRedraw = True
    iPic.BackColor = nGrid.BackColorFixed
    iPic.ForeColor = nGrid.GridColorFixed
    iPic.ScaleMode = vbPixels
    
    iPic.Line (1, 8)-(15, 8)
    iPic.Line (1, 8)-(8, 1)
    iPic.Line (2, 8)-(8, 2)
    iPic.Line (8, 2)-(14, 8)
    iPic.Line (7, 2)-(13, 8)
    Set iImage = iPic.Image
    
    mGridsArrowUpImageCollection.Add iImage, CStr(nGrid.BackColorFixed)

    iPic.Cls
    Set iPic.Picture = Nothing
    iPic.PaintPicture iImage, 0, 0, iPic.ScaleWidth, iPic.ScaleHeight, 0, iPic.ScaleHeight, iPic.ScaleWidth, -iPic.ScaleHeight, vbSrcCopy
    Set iImage = iPic.Image
    mGridsArrowDownImageCollection.Add iImage, CStr(nGrid.BackColorFixed)
    
    On Error Resume Next
    nGrid.Parent.Controls.Remove "picAux1x"

TheExit:
End Sub

Public Function GetGridReportStyleID(nGridReportStyle As GridReportStyle, nNumberNewCustomStyle As Long) As String
    Dim c As Long
    Dim iGridReportStyle As GridReportStyle
    Dim iStyleID As String
    Dim iSt
    
    c = 1
    Set iGridReportStyle = GetGridReportStyle("GRStyle" & c)
    Do Until iGridReportStyle.Tag = ""
        If GridReportStylesAreEqual(iGridReportStyle, nGridReportStyle) Then
            iStyleID = iGridReportStyle.Tag
            Exit Do
        End If
        c = c + 1
        Set iGridReportStyle = GetGridReportStyle("GRStyle" & c)
    Loop
    
    If iStyleID = "" Then
        For Each iSt In gGridReportStyles
            Set iGridReportStyle = iSt
            If GridReportStylesAreEqual(iGridReportStyle, nGridReportStyle) Then
                iStyleID = iGridReportStyle.Tag
                Exit For
            End If
        Next
    End If
    
    If iStyleID = "" Then
        c = 1
        Set iGridReportStyle = GetGridReportStyle("Custom" & c)
        Do Until iGridReportStyle.Tag = ""
            If GridReportStylesAreEqual(iGridReportStyle, nGridReportStyle) Then
                iStyleID = iGridReportStyle.Tag
                Exit Do
            End If
            c = c + 1
            Set iGridReportStyle = GetGridReportStyle("Custom" & c)
        Loop
    End If
    
    nNumberNewCustomStyle = c
    GetGridReportStyleID = iStyleID
End Function

Public Function GridReportStylesAreEqual(nGridReportStyle1 As GridReportStyle, nGridReportStyle2 As GridReportStyle) As Boolean
    Dim iChange As Boolean
    
    If nGridReportStyle1.LineWidth <> nGridReportStyle2.LineWidth Then
        iChange = True
    ElseIf nGridReportStyle1.PrintHeadersBackground <> nGridReportStyle2.PrintHeadersBackground Then
        iChange = True
    ElseIf nGridReportStyle1.HeadersBackgroundColor <> nGridReportStyle2.HeadersBackgroundColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintFixedColsBackground <> nGridReportStyle2.PrintFixedColsBackground Then
        iChange = True
    ElseIf nGridReportStyle1.PrintOtherBackgrounds <> nGridReportStyle2.PrintOtherBackgrounds Then
        iChange = True
    ElseIf nGridReportStyle1.PrintOuterBorder <> nGridReportStyle2.PrintOuterBorder Then
        iChange = True
    ElseIf nGridReportStyle1.OuterBorderColor <> nGridReportStyle2.OuterBorderColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintHeadersBorder <> nGridReportStyle2.PrintHeadersBorder Then
        iChange = True
    ElseIf nGridReportStyle1.HeadersBorderColor <> nGridReportStyle2.HeadersBorderColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintColumnsDataLines <> nGridReportStyle2.PrintColumnsDataLines Then
        iChange = True
    ElseIf nGridReportStyle1.ColumnsDataLinesColor <> nGridReportStyle2.ColumnsDataLinesColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintColumnsHeadersLines <> nGridReportStyle2.PrintColumnsHeadersLines Then
        iChange = True
    ElseIf nGridReportStyle1.ColumnsHeadersLinesColor <> nGridReportStyle2.ColumnsHeadersLinesColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintRowsLines <> nGridReportStyle2.PrintRowsLines Then
        iChange = True
    ElseIf nGridReportStyle1.RowsLinesColor <> nGridReportStyle2.RowsLinesColor Then
        iChange = True
    ElseIf nGridReportStyle1.PrintHeadersSeparatorLine <> nGridReportStyle2.PrintHeadersSeparatorLine Then
        iChange = True
    ElseIf nGridReportStyle1.LineWidthHeadersSeparatorLine <> nGridReportStyle2.LineWidthHeadersSeparatorLine Then
        iChange = True
    End If
    
    GridReportStylesAreEqual = Not iChange
End Function

Public Function GetGridReportStyle(nStyleID As String) As GridReportStyle
    Dim iSt
    
    Set GetGridReportStyle = New GridReportStyle
    
    Select Case nStyleID
        Case "GRStyle1"
            GetGridReportStyle.Tag = "GRStyle1"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(239, 239, 239)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = True
            GetGridReportStyle.OuterBorderColor = RGB(31, 31, 31)
            GetGridReportStyle.PrintHeadersBorder = False
            GetGridReportStyle.HeadersBorderColor = RGB(207, 207, 207)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(207, 207, 207)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(207, 207, 207)
            GetGridReportStyle.PrintRowsLines = True
            GetGridReportStyle.RowsLinesColor = RGB(207, 207, 207)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
        
        Case "GRStyle2"
            GetGridReportStyle.Tag = "GRStyle2"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(244, 249, 255)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = False
            GetGridReportStyle.OuterBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintHeadersBorder = True
            GetGridReportStyle.HeadersBorderColor = RGB(29, 4, 145)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintRowsLines = False
            GetGridReportStyle.RowsLinesColor = RGB(230, 224, 254)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
        
        Case "GRStyle3"
            GetGridReportStyle.Tag = "GRStyle3"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(244, 249, 255)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = False
            GetGridReportStyle.OuterBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintHeadersBorder = False
            GetGridReportStyle.HeadersBorderColor = RGB(29, 4, 145)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintRowsLines = False
            GetGridReportStyle.RowsLinesColor = RGB(230, 224, 254)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 10
        
        Case "GRStyle4"
            GetGridReportStyle.Tag = "GRStyle4"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(252, 243, 243)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = False
            GetGridReportStyle.OuterBorderColor = RGB(139, 40, 35)
            GetGridReportStyle.PrintHeadersBorder = True
            GetGridReportStyle.HeadersBorderColor = RGB(139, 40, 35)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(237, 197, 194)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(237, 197, 194)
            GetGridReportStyle.PrintRowsLines = False
            GetGridReportStyle.RowsLinesColor = RGB(247, 230, 227)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
        
        Case "GRStyle5"
            GetGridReportStyle.Tag = "GRStyle5"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(244, 249, 255)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = True
            GetGridReportStyle.OuterBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintHeadersBorder = False
            GetGridReportStyle.HeadersBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintRowsLines = False
            GetGridReportStyle.RowsLinesColor = RGB(230, 224, 254)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
        
        Case "GRStyle6"
            GetGridReportStyle.Tag = "GRStyle6"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(244, 249, 255)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = True
            GetGridReportStyle.OuterBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintHeadersBorder = False
            GetGridReportStyle.HeadersBorderColor = RGB(40, 5, 207)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(177, 162, 253)
            GetGridReportStyle.PrintRowsLines = True
            GetGridReportStyle.RowsLinesColor = RGB(230, 224, 254)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
            
        Case "GRStyle7"
            GetGridReportStyle.Tag = "GRStyle7"
            
            GetGridReportStyle.LineWidth = 3
            GetGridReportStyle.PrintHeadersBackground = True
            GetGridReportStyle.HeadersBackgroundColor = RGB(252, 243, 243)
            GetGridReportStyle.PrintFixedColsBackground = False
            GetGridReportStyle.PrintOtherBackgrounds = True
            
            GetGridReportStyle.PrintOuterBorder = True
            GetGridReportStyle.OuterBorderColor = RGB(139, 40, 35)
            GetGridReportStyle.PrintHeadersBorder = False
            GetGridReportStyle.HeadersBorderColor = RGB(139, 40, 35)
            GetGridReportStyle.PrintColumnsDataLines = True
            GetGridReportStyle.ColumnsDataLinesColor = RGB(237, 197, 194)
            GetGridReportStyle.PrintColumnsHeadersLines = True
            GetGridReportStyle.ColumnsHeadersLinesColor = RGB(237, 197, 194)
            GetGridReportStyle.PrintRowsLines = True
            GetGridReportStyle.RowsLinesColor = RGB(247, 230, 227)
            GetGridReportStyle.PrintHeadersSeparatorLine = True
            GetGridReportStyle.LineWidthHeadersSeparatorLine = 3
        
        Case Else
            If Left(nStyleID, 12) = "GRStyleAdded" Then
'                Set GetGridReportStyle = gGridReportStyles(Val(Mid(nStyleID, 13))).Clone
                
                For Each iSt In gGridReportStyles
                    If iSt.Tag = nStyleID Then
                        Set GetGridReportStyle = iSt.Clone
                        Exit For
                    End If
                Next
            End If
            If Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintOuterBorder", -44)) <> -44 Then
                GetGridReportStyle.Tag = nStyleID
                
                GetGridReportStyle.PrintOuterBorder = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintOuterBorder", -1))
                GetGridReportStyle.PrintHeadersBorder = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintHeadersBorder", -1))
                GetGridReportStyle.PrintColumnsDataLines = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintColumnsDataLines", -1))
                GetGridReportStyle.PrintColumnsHeadersLines = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintColumnsHeadersLines", -1))
                GetGridReportStyle.PrintRowsLines = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintRowsLines", -1))
                GetGridReportStyle.PrintHeadersSeparatorLine = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintHeadersSeparatorLine", -1))
                GetGridReportStyle.PrintHeadersBackground = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintHeadersBackground", -1))
                GetGridReportStyle.PrintFixedColsBackground = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintFixedColsBackground", 0))
                GetGridReportStyle.PrintOtherBackgrounds = CBool(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_PrintOtherBackgrounds", -1))
                GetGridReportStyle.LineWidth = GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_LineWidth", 3)
                GetGridReportStyle.LineWidthHeadersSeparatorLine = GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_LineWidthHeadersSeparatorLine", 3)
            
                GetGridReportStyle.OuterBorderColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_OuterBorderColor", RGB(31, 31, 31)))
                GetGridReportStyle.ColumnsDataLinesColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_ColumnsDataLinesColor", RGB(207, 207, 207)))
                GetGridReportStyle.ColumnsHeadersLinesColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_ColumnsHeadersLinesColor", RGB(207, 207, 207)))
                GetGridReportStyle.RowsLinesColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_RowsLinesColor", RGB(207, 207, 207)))
                GetGridReportStyle.HeadersBorderColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_HeadersBorderColor", RGB(207, 207, 207)))
                GetGridReportStyle.HeadersBackgroundColor = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & nStyleID & "_HeadersBackgroundColor", RGB(239, 239, 239)))
            Else
                GetGridReportStyle.Tag = ""
            End If
    End Select
End Function


Public Function GetIDGridReportStyleSaved(nGridReportStyle As GridReportStyle) As String
    Dim iStyleID As String
    Dim iNumberNewCustomStyle As Long
    
    iStyleID = GetGridReportStyleID(nGridReportStyle, iNumberNewCustomStyle)
    
    If iStyleID = "" Then
        If iNumberNewCustomStyle > 10 Then
            iNumberNewCustomStyle = 1
            iNumberNewCustomStyle = Val(GetSetting(AppNameForRegistry, "PrintingSettings", "LastCustomGridReportStyleNumber", 0)) + 1
            If iNumberNewCustomStyle > 10 Then
                iNumberNewCustomStyle = 1
            End If
            SaveSetting AppNameForRegistry, "PrintingSettings", "LastCustomGridReportStyleNumber", iNumberNewCustomStyle
        End If
        iStyleID = "Custom" & iNumberNewCustomStyle
        
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_LineWidth", nGridReportStyle.LineWidth
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintHeadersBackground", CLng(nGridReportStyle.PrintHeadersBackground)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_HeadersBackgroundColor", nGridReportStyle.HeadersBackgroundColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintFixedColsBackground", CLng(nGridReportStyle.PrintFixedColsBackground)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintOtherBackgrounds", CLng(nGridReportStyle.PrintOtherBackgrounds)
        
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintOuterBorder", CLng(nGridReportStyle.PrintOuterBorder)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_OuterBorderColor", nGridReportStyle.OuterBorderColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintHeadersBorder", CLng(nGridReportStyle.PrintHeadersBorder)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_HeadersBorderColor", nGridReportStyle.HeadersBorderColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintColumnsDataLines", CLng(nGridReportStyle.PrintColumnsDataLines)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_ColumnsDataLinesColor", nGridReportStyle.ColumnsDataLinesColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintColumnsHeadersLines", CLng(nGridReportStyle.PrintColumnsHeadersLines)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_ColumnsHeadersLinesColor", nGridReportStyle.ColumnsHeadersLinesColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintRowsLines", CLng(nGridReportStyle.PrintRowsLines)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_RowsLinesColor", nGridReportStyle.RowsLinesColor
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_PrintHeadersSeparatorLine", CLng(nGridReportStyle.PrintHeadersSeparatorLine)
        SaveSetting AppNameForRegistry, "PrintingSettings", "GridReportStyle:" & iStyleID & "_LineWidthHeadersSeparatorLine", nGridReportStyle.LineWidthHeadersSeparatorLine
        
    End If
    
    GetIDGridReportStyleSaved = iStyleID
    
End Function

